File: client/ts/src/framer/codec.ts
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { type WebsocketMessage } from "@synnaxlabs/freighter";
import {
  binary,
  DataType,
  type SeriesPayload,
  TimeRange,
  TimeStamp,
} from "@synnaxlabs/x";
import { type z } from "zod";

import { type channel } from "@/channel";
import { ValidationError } from "@/errors";
import { type Frame, type Payload } from "@/framer/frame";
import { WriterCommand } from "@/framer/payload";
import { type StreamerResponse } from "@/framer/streamer";
import { type WriteRequest } from "@/framer/writer";

const seriesPldLength = (series: SeriesPayload): number =>
  series.data.byteLength / series.dataType.density.valueOf();

interface KeyedSeries extends SeriesPayload {
  key: number;
}

const sortFramePayloadByKey = (framePayload: Payload): void => {
  const { keys, series } = framePayload;
  keys.forEach((key, index) => {
    (series[index] as KeyedSeries).key = key;
  });
  series.sort((a, b) => (a as KeyedSeries).key - (b as KeyedSeries).key);
  keys.sort((a, b) => a - b);
  // @ts-expect-error - deleting static property keys.
  series.forEach((ser) => delete (ser as KeyedSeries).key);
};

const ZERO_ALIGNMENTS_FLAG_POS = 5;
const EQUAL_ALIGNMENTS_FLAG_POS = 4;
const EQUAL_LENGTHS_FLAG_POS = 3;
const EQUAL_TIME_RANGES_FLAG_POS = 2;
const TIME_RANGES_ZERO_FLAG_POS = 1;
const ALL_CHANNELS_PRESENT_FLAG_POS = 0;

const TIMESTAMP_SIZE = DataType.TIMESTAMP.density.valueOf();
const ALIGNMENT_SIZE = 8;
const DATA_LENGTH_SIZE = 4;
const KEY_SIZE = 4;
const SEQ_NUM_SIZE = 4;
const FLAGS_SIZE = 1;

interface CodecState {
  keys: channel.Keys;
  keyDataTypes: Map<channel.Key, DataType>;
  hasVariableDataTypes: boolean;
}

export class Codec {
  contentType: string = "application/sy-framer";
  private states: Map<number, CodecState> = new Map();
  private currState: CodecState | undefined;
  private seqNum: number = 0;

  constructor(keys: channel.Keys = [], dataTypes: DataType[] = []) {
    if (keys.length > 0 || dataTypes.length > 0) this.update(keys, dataTypes);
  }

  update(keys: channel.Keys, dataTypes: DataType[]): void {
    this.seqNum++;
    const state = {
      keys,
      keyDataTypes: new Map(),
      hasVariableDataTypes: false,
    };
    keys.forEach((k, i) => {
      const dt = dataTypes[i];
      state.keyDataTypes.set(k, dt);
      if (dt.isVariable) state.hasVariableDataTypes = true;
    });
    state.keys.sort();
    this.states.set(this.seqNum, state);
    this.currState = state;
  }

  private throwIfNotUpdated(op: string): void {
    if (this.seqNum < 1)
      throw new ValidationError(`
      The codec has not been updated with a list of channels and data types.
      Please call the update method before calling ${op}.
      `);
  }

  encode(payload: unknown, startOffset: number = 0): Uint8Array {
    this.throwIfNotUpdated("encode");
    let src = payload as Payload;
    if (payload != null && typeof payload === "object" && "toPayload" in payload)
      src = (payload as Frame).toPayload();
    sortFramePayloadByKey(src);
    let currDataSize = -1;
    let startTime: TimeStamp | undefined;
    let endTime: TimeStamp | undefined;
    let currAlignment: bigint | undefined;
    let byteArraySize = startOffset + FLAGS_SIZE + SEQ_NUM_SIZE;
    let equalLengthsFlag = !this.currState?.hasVariableDataTypes;
    let equalTimeRangesFlag = true;
    let timeRangesZeroFlag = true;
    let channelFlag = true;
    let equalAlignmentsFlag = true;
    let zeroAlignmentsFlag = true;

    if (src.keys.length !== this.currState?.keys.length) {
      channelFlag = false;
      byteArraySize += src.keys.length * KEY_SIZE;
    }

    src.series.forEach((series, i) => {
      const pldLength = seriesPldLength(series);
      const key = src.keys[i];
      const dt = this.currState?.keyDataTypes.get(key);
      if (dt == null)
        throw new ValidationError(
          `Channel ${key} was not provided in the list of channels when opening the writer`,
        );
      if (!dt.equals(series.dataType))
        throw new ValidationError(
          `Series data type of ${series.dataType.toString()} does not match the data type of ${dt.toString()} for channel ${key}`,
        );

      byteArraySize += series.data.byteLength;
      if (currDataSize === -1) {
        currDataSize = pldLength;
        startTime = series.timeRange?.start;
        endTime = series.timeRange?.end;
        currAlignment = BigInt(series.alignment ?? 0n);
        return;
      }
      if (currDataSize !== pldLength) equalLengthsFlag = false;
      if (
        startTime?.valueOf() !== series.timeRange?.start.valueOf() ||
        endTime?.valueOf() !== series.timeRange?.end.valueOf()
      )
        equalTimeRangesFlag = false;
      if (currAlignment !== BigInt(series.alignment ?? 0)) equalAlignmentsFlag = false;
    });

    timeRangesZeroFlag = equalTimeRangesFlag && startTime == null && endTime == null;

    zeroAlignmentsFlag =
      equalAlignmentsFlag && (currAlignment === undefined || currAlignment === 0n);

    if (equalLengthsFlag) byteArraySize += DATA_LENGTH_SIZE;
    else byteArraySize += src.keys.length * DATA_LENGTH_SIZE;

    if (!timeRangesZeroFlag)
      if (equalTimeRangesFlag) byteArraySize += TIMESTAMP_SIZE * 2;
      else byteArraySize += src.keys.length * TIMESTAMP_SIZE * 2;

    if (!zeroAlignmentsFlag)
      if (equalAlignmentsFlag) byteArraySize += ALIGNMENT_SIZE;
      else byteArraySize += src.keys.length * ALIGNMENT_SIZE;

    const buffer = new Uint8Array(byteArraySize);
    const view = new DataView(buffer.buffer);
    let offset = startOffset;
    buffer[offset] =
      (Number(zeroAlignmentsFlag) << ZERO_ALIGNMENTS_FLAG_POS) |
      (Number(equalAlignmentsFlag) << EQUAL_ALIGNMENTS_FLAG_POS) |
      (Number(equalLengthsFlag) << EQUAL_LENGTHS_FLAG_POS) |
      (Number(equalTimeRangesFlag) << EQUAL_TIME_RANGES_FLAG_POS) |
      (Number(timeRangesZeroFlag) << TIME_RANGES_ZERO_FLAG_POS) |
      (Number(channelFlag) << ALL_CHANNELS_PRESENT_FLAG_POS);
    offset++;
    view.setUint32(offset, this.seqNum, true);
    offset += SEQ_NUM_SIZE;

    if (equalLengthsFlag) {
      view.setUint32(offset, currDataSize, true);
      offset += DATA_LENGTH_SIZE;
    }

    if (equalTimeRangesFlag && !timeRangesZeroFlag) {
      view.setBigUint64(offset, startTime?.valueOf() ?? 0n, true);
      view.setBigUint64(offset, endTime?.valueOf() ?? 0n, true);
      offset += TIMESTAMP_SIZE * 2;
    }

    if (equalAlignmentsFlag && !zeroAlignmentsFlag) {
      view.setBigUint64(offset, currAlignment ?? 0n, true);
      offset += ALIGNMENT_SIZE;
    }

    src.series.forEach((series, i) => {
      if (!channelFlag) {
        view.setUint32(offset, src.keys[i], true);
        offset += KEY_SIZE;
      }
      if (!equalLengthsFlag) {
        let seriesLengthOrSize = series.data.byteLength;
        if (!series.dataType.isVariable) seriesLengthOrSize = seriesPldLength(series);
        view.setUint32(offset, seriesLengthOrSize, true);
        offset += DATA_LENGTH_SIZE;
      }
      buffer.set(new Uint8Array(series.data), offset);
      offset += series.data.byteLength;
      if (!equalTimeRangesFlag && !timeRangesZeroFlag) {
        view.setBigUint64(offset, series.timeRange?.start.valueOf() ?? 0n, true);
        view.setBigUint64(offset, series.timeRange?.end.valueOf() ?? 0n, true);
        offset += TIMESTAMP_SIZE * 2;
      }
      if (!equalAlignmentsFlag && !zeroAlignmentsFlag) {
        view.setBigUint64(offset, series.alignment ?? 0n, true);
        offset += ALIGNMENT_SIZE;
      }
    });
    return buffer;
  }

  decode(data: Uint8Array | ArrayBuffer, offset: number = 0): Payload {
    this.throwIfNotUpdated("decode");
    const src = data instanceof Uint8Array ? data : new Uint8Array(data);
    const returnFrame: Payload = { keys: [], series: [] };
    let index = offset;
    let sizeRepresentation = 0;
    let currSize = 0;
    let startTime: TimeStamp | undefined;
    let endTime: TimeStamp | undefined;
    let currAlignment: bigint | undefined;

    const view = new DataView(src.buffer, src.byteOffset, src.byteLength);
    const zeroAlignmentsFlag = Boolean((src[index] >> ZERO_ALIGNMENTS_FLAG_POS) & 1);
    const equalAlignmentsFlag = Boolean((src[index] >> EQUAL_ALIGNMENTS_FLAG_POS) & 1);
    const sizeFlag = Boolean((src[index] >> EQUAL_LENGTHS_FLAG_POS) & 1);
    const equalTimeRangesFlag = Boolean((src[index] >> EQUAL_TIME_RANGES_FLAG_POS) & 1);
    const timeRangesZeroFlag = Boolean((src[index] >> TIME_RANGES_ZERO_FLAG_POS) & 1);
    const channelFlag = Boolean((src[index] >> ALL_CHANNELS_PRESENT_FLAG_POS) & 1);
    index++;

    const seqNum = view.getUint32(index, true);
    index += SEQ_NUM_SIZE;
    const state = this.states.get(seqNum);
    if (state == null) return returnFrame;

    if (sizeFlag) {
      if (index + DATA_LENGTH_SIZE > view.byteLength) return returnFrame;
      sizeRepresentation = view.getUint32(index, true);
      index += DATA_LENGTH_SIZE;
    }

    if (equalTimeRangesFlag && !timeRangesZeroFlag) {
      if (index + TIMESTAMP_SIZE > view.byteLength) return returnFrame;
      startTime = new TimeStamp(view.getBigUint64(index, true));
      index += TIMESTAMP_SIZE;
      endTime = new TimeStamp(view.getBigUint64(index, true));
      index += TIMESTAMP_SIZE;
    }

    if (equalAlignmentsFlag && !zeroAlignmentsFlag) {
      if (index + ALIGNMENT_SIZE > view.byteLength) return returnFrame;
      currAlignment = view.getBigUint64(index, true);
      index += ALIGNMENT_SIZE;
    }

    if (channelFlag) returnFrame.keys = [...state.keys];
    state.keys.forEach((k, i) => {
      if (!channelFlag) {
        if (index >= view.byteLength) return;
        const frameKey = view.getUint32(index, true);
        if (frameKey !== k) return;
        index += KEY_SIZE;
        returnFrame.keys.push(k);
      }
      const dataType = state.keyDataTypes.get(k) as DataType;
      currSize = 0;
      if (!sizeFlag) {
        if (index + DATA_LENGTH_SIZE > view.byteLength) return;
        currSize = view.getUint32(index, true);
        index += DATA_LENGTH_SIZE;
      } else currSize = sizeRepresentation;

      let dataByteLength = currSize;
      if (!dataType.isVariable) dataByteLength *= dataType.density.valueOf();
      if (index + dataByteLength > view.byteLength) {
        returnFrame.keys.splice(i, 1);
        return;
      }
      const currSeries: SeriesPayload = {
        dataType,
        data: src.slice(index, index + dataByteLength).buffer,
      };
      index += dataByteLength;
      if (!equalTimeRangesFlag && !timeRangesZeroFlag) {
        if (index + TIMESTAMP_SIZE * 2 > view.byteLength) return;
        const start = view.getBigUint64(index, true);
        index += TIMESTAMP_SIZE;
        const end = view.getBigUint64(index, true);
        index += TIMESTAMP_SIZE;
        currSeries.timeRange = new TimeRange({ start, end });
      } else if (!timeRangesZeroFlag)
        currSeries.timeRange = new TimeRange({
          start: startTime?.valueOf() ?? 0n,
          end: endTime?.valueOf() ?? 0n,
        });
      else currSeries.timeRange = new TimeRange({ start: 0n, end: 0n });

      if (!equalAlignmentsFlag && !zeroAlignmentsFlag) {
        if (index + ALIGNMENT_SIZE > view.byteLength) return;
        currAlignment = view.getBigUint64(index, true);
        index += ALIGNMENT_SIZE;
        currSeries.alignment = currAlignment;
      } else if (!zeroAlignmentsFlag) currSeries.alignment = currAlignment;
      else currSeries.alignment = 0n;

      returnFrame.series.push(currSeries);
    });
    return returnFrame;
  }
}

export const LOW_PER_SPECIAL_CHAR = 254;
const LOW_PERF_SPECIAL_CHAR_BUF = new Uint8Array([LOW_PER_SPECIAL_CHAR]);
export const HIGH_PERF_SPECIAL_CHAR = 255;
const HIGH_PERF_SPECIAL_CHAR_BUF = new Uint8Array([HIGH_PERF_SPECIAL_CHAR]);
const CONTENT_TYPE = "application/sy-framer";

export class WSWriterCodec implements binary.Codec {
  contentType = CONTENT_TYPE;
  private base: Codec;
  private lowPerfCodec: binary.Codec;

  constructor(base: Codec) {
    this.base = base;
    this.lowPerfCodec = binary.JSON_CODEC;
  }

  encode(payload: unknown): Uint8Array {
    const pld = payload as WebsocketMessage<WriteRequest>;
    if (pld.type == "close" || pld.payload?.command != WriterCommand.Write) {
      const data = this.lowPerfCodec.encode(pld);
      const b = new Uint8Array({ length: data.byteLength + 1 });
      b.set(LOW_PERF_SPECIAL_CHAR_BUF, 0);
      b.set(new Uint8Array(data), 1);
      return b;
    }
    const data = this.base.encode(pld.payload?.frame, 1);
    data.set(HIGH_PERF_SPECIAL_CHAR_BUF, 0);
    return data;
  }

  decode<P extends z.ZodType>(data: Uint8Array | ArrayBuffer, schema?: P): z.infer<P> {
    const dv = new DataView(data instanceof Uint8Array ? data.buffer : data);
    const codec = dv.getUint8(0);
    if (codec === LOW_PER_SPECIAL_CHAR)
      return this.lowPerfCodec.decode(data.slice(1), schema);
    const v: WebsocketMessage<WriteRequest> = { type: "data" };
    const frame = this.base.decode(data, 1);
    v.payload = { command: WriterCommand.Write, frame };
    return v as z.infer<P>;
  }
}

export class WSStreamerCodec implements binary.Codec {
  contentType = CONTENT_TYPE;
  private base: Codec;
  private lowPerfCodec: binary.Codec;

  constructor(base: Codec) {
    this.base = base;
    this.lowPerfCodec = binary.JSON_CODEC;
  }

  encode(payload: unknown): Uint8Array {
    return this.lowPerfCodec.encode(payload);
  }

  decode<P extends z.ZodType>(data: Uint8Array | ArrayBuffer, schema?: P): z.infer<P> {
    const dv = new DataView(data instanceof Uint8Array ? data.buffer : data);
    const codec = dv.getUint8(0);
    if (codec === LOW_PER_SPECIAL_CHAR)
      return this.lowPerfCodec.decode(data.slice(1), schema);
    const v: WebsocketMessage<StreamerResponse> = {
      type: "data",
      payload: { frame: this.base.decode(data, 1) },
    };
    return v as z.infer<P>;
  }
}

File: client/cpp/framer/codec.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include <algorithm>
#include <memory>
#include <sstream>
#include <vector>

#include "client/cpp/framer/framer.h"

namespace synnax::framer {
uint8_t CodecFlags::encode() const {
    uint8_t b = 0;
    b = x::binary::set_bit(b, FlagPosition::EqualLengths, equal_lens);
    b = x::binary::set_bit(b, FlagPosition::EqualTimeRanges, equal_time_ranges);
    b = x::binary::set_bit(b, FlagPosition::TimeRangesZero, time_ranges_zero);
    b = x::binary::set_bit(b, FlagPosition::AllChannelsPresent, all_channels_present);
    b = x::binary::set_bit(b, FlagPosition::EqualAlignments, equal_alignments);
    b = x::binary::set_bit(b, FlagPosition::ZeroAlignments, zero_alignments);
    return b;
}

CodecFlags CodecFlags::decode(const uint8_t b) {
    CodecFlags f;
    f.equal_lens = x::binary::get_bit(b, FlagPosition::EqualLengths);
    f.equal_time_ranges = x::binary::get_bit(b, FlagPosition::EqualTimeRanges);
    f.time_ranges_zero = x::binary::get_bit(b, FlagPosition::TimeRangesZero);
    f.all_channels_present = x::binary::get_bit(b, FlagPosition::AllChannelsPresent);
    f.equal_alignments = x::binary::get_bit(b, FlagPosition::EqualAlignments);
    f.zero_alignments = x::binary::get_bit(b, FlagPosition::ZeroAlignments);
    return f;
}

x::errors::Error Codec::update(const std::vector<channel::Key> &keys) {
    this->seq_num++;
    auto [channels, err] = this->channel_client.retrieve(keys);
    if (err) return err;
    Codec::State state;
    for (const auto &ch: channels) {
        state.keys.insert(ch.key);
        state.key_data_types[ch.key] = ch.data_type;
        if (ch.data_type.is_variable()) state.has_variable_data_types = true;
    }
    this->states[seq_num] = state;
    return x::errors::NIL;
}

void Codec::throw_if_uninitialized() const {
    if (this->seq_num < 1) throw std::runtime_error("codec is uninitialized");
}

Codec::Codec(
    const std::vector<channel::Key> &channels,
    const std::vector<x::telem::DataType> &data_types
):
    seq_num(1), channel_client(nullptr, nullptr) {
    Codec::State state;
    state.key_data_types.reserve(channels.size());
    for (size_t i = 0; i < channels.size(); i++) {
        auto k = channels[i];
        const auto &dt = data_types[i];
        state.keys.insert(k);
        state.key_data_types[k] = dt;
        if (dt.is_variable()) state.has_variable_data_types = true;
    }
    this->states[this->seq_num] = state;
}

constexpr std::size_t ALIGNMENT_SIZE = 8;
constexpr std::size_t DATA_LENGTH_SIZE = 4;
constexpr std::size_t KEY_SIZE = 4;
constexpr std::size_t FLAGS_SIZE = 1;
constexpr std::size_t SEQ_NUM_SIZE = 4;
constexpr std::size_t TIME_RANGE_SIZE = 16;

x::errors::Error
Codec::encode(const x::telem::Frame &frame, std::vector<uint8_t> &output) {
    this->throw_if_uninitialized();
    CodecFlags flags;
    size_t byte_array_size = FLAGS_SIZE + SEQ_NUM_SIZE;

    auto state = this->states[this->seq_num];

    if (frame.channels->size() != state.keys.size()) {
        flags.all_channels_present = false;
        byte_array_size += frame.channels->size() * KEY_SIZE;
    }

    this->sorting_indices.resize(frame.size());
    for (size_t i = 0; i < frame.channels->size(); i++) {
        auto k = frame.channels->at(i);
        auto &ser = frame.series->at(i);
        auto dt = state.key_data_types.find(k);
        if (dt == state.key_data_types.end())
            return x::errors::Error(
                x::errors::VALIDATION,
                "frame contains extra key " + std::to_string(k) +
                    "not provided when opening the writer"
            );
        if (dt->second != ser.data_type())
            return x::errors::Error(
                x::errors::VALIDATION,
                "data type " + dt->second + " for channel + " + std::to_string(k) +
                    " does not match series data type " + ser.data_type()
            );
        this->sorting_indices[i] = {k, i};
    }
    std::sort(sorting_indices.begin(), sorting_indices.end());

    flags.equal_lens = !state.has_variable_data_types;
    size_t cur_data_size = 0;
    bool first_series = true;
    x::telem::TimeRange ref_tr = {};
    x::telem::Alignment ref_alignment;

    for (const auto &[key, idx]: sorting_indices) {
        const x::telem::Series &series = frame.series->at(idx);
        byte_array_size += series.byte_size();
        if (first_series) {
            cur_data_size = series.size();
            ref_tr = series.time_range;
            ref_alignment = series.alignment;
            first_series = false;
            continue;
        }
        if (series.size() != cur_data_size) flags.equal_lens = false;
        if (series.time_range != ref_tr) flags.equal_time_ranges = false;
        if (series.alignment != ref_alignment) flags.equal_alignments = false;
    }

    flags.time_ranges_zero = flags.equal_time_ranges &&
                             ref_tr.start.nanoseconds() == 0 &&
                             ref_tr.end.nanoseconds() == 0;

    flags.zero_alignments = flags.equal_alignments && ref_alignment == 0;

    if (!flags.equal_lens)
        byte_array_size += frame.channels->size() * DATA_LENGTH_SIZE;
    else
        byte_array_size += DATA_LENGTH_SIZE;

    if (!flags.time_ranges_zero) {
        if (!flags.equal_time_ranges)
            byte_array_size += frame.channels->size() * TIME_RANGE_SIZE;
        else
            byte_array_size += TIME_RANGE_SIZE;
    }

    if (!flags.zero_alignments) {
        if (!flags.equal_alignments)
            byte_array_size += frame.channels->size() * ALIGNMENT_SIZE;
        else
            byte_array_size += ALIGNMENT_SIZE;
    }

    x::binary::Writer buf(output, byte_array_size);

    if (buf.uint8(flags.encode()) != 1)
        return x::errors::Error(x::errors::UNEXPECTED, "failed to write flags");
    if (buf.uint32(this->seq_num) != 4)
        return x::errors::Error(
            x::errors::UNEXPECTED,
            "failed to write sequence number"
        );

    if (flags.equal_lens) {
        if (buf.uint32(static_cast<uint32_t>(cur_data_size)) != 4)
            return x::errors::Error(
                x::errors::UNEXPECTED,
                "failed to write data length"
            );
    }

    if (flags.equal_time_ranges && !flags.time_ranges_zero) {
        if (buf.int64(ref_tr.start.nanoseconds()) != 8)
            return x::errors::Error(
                x::errors::UNEXPECTED,
                "failed to write time range start"
            );
        if (buf.int64(ref_tr.end.nanoseconds()) != 8)
            return x::errors::Error(
                x::errors::UNEXPECTED,
                "failed to write time range end"
            );
    }

    if (flags.equal_alignments && !flags.zero_alignments) {
        if (buf.uint64(ref_alignment.uint64()) != 8)
            return x::errors::Error(x::errors::UNEXPECTED, "failed to write alignment");
    }

    for (const auto &[key, idx]: sorting_indices) {
        const x::telem::Series &ser = frame.series->at(idx);
        const auto byte_size = ser.byte_size();

        if (!flags.all_channels_present) {
            if (buf.uint32(key) != 4)
                return x::errors::Error(
                    x::errors::UNEXPECTED,
                    "failed to write channel key"
                );
        }

        if (!flags.equal_lens) {
            const auto size = ser.data_type().is_variable() ? byte_size : ser.size();
            if (buf.uint32(static_cast<uint32_t>(size)) != 4)
                return x::errors::Error(
                    x::errors::UNEXPECTED,
                    "failed to write series length"
                );
        }

        if (buf.write(ser.data(), byte_size) != byte_size)
            return x::errors::Error(
                x::errors::UNEXPECTED,
                "failed to write series data: expected " + std::to_string(byte_size) +
                    " bytes"
            );

        if (!flags.equal_time_ranges) {
            if (buf.int64(ser.time_range.start.nanoseconds()) != 8)
                return x::errors::Error(
                    x::errors::UNEXPECTED,
                    "failed to write series time range start"
                );
            if (buf.int64(ser.time_range.end.nanoseconds()) != 8)
                return x::errors::Error(
                    x::errors::UNEXPECTED,
                    "failed to write series time range end"
                );
        }

        if (!flags.equal_alignments) {
            if (buf.uint64(ser.alignment.uint64()) != 8)
                return x::errors::Error(
                    x::errors::UNEXPECTED,
                    "failed to write series alignment"
                );
        }
    }

    return x::errors::NIL;
}

std::pair<x::telem::Frame, x::errors::Error>
Codec::decode(const std::vector<uint8_t> &data) const {
    return this->decode(data.data(), data.size());
}

std::pair<x::telem::Frame, x::errors::Error>
Codec::decode(const uint8_t *data, const size_t size) const {
    this->throw_if_uninitialized();
    auto reader = x::binary::Reader(data, size);
    x::telem::Frame frame;
    uint32_t data_len = 0;
    x::telem::TimeRange ref_tr = {};
    x::telem::Alignment ref_alignment;
    auto flags = CodecFlags::decode(reader.uint8());

    auto seq_num = reader.uint32();
    auto state = this->states.at(seq_num);

    if (flags.equal_lens) data_len = reader.uint32();

    if (flags.equal_time_ranges && !flags.time_ranges_zero) {
        ref_tr.start = x::telem::TimeStamp(reader.int64());
        ref_tr.end = x::telem::TimeStamp(reader.int64());
    } else if (flags.time_ranges_zero)
        ref_tr = x::telem::TimeRange{x::telem::TimeStamp(0), x::telem::TimeStamp(0)};

    if (flags.equal_alignments && !flags.zero_alignments)
        ref_alignment = x::telem::Alignment(reader.uint64());

    auto decode_series = [&](const channel::Key key) {
        // when the series is a variable data type, we use its byte capacity instead
        // of its length.
        uint32_t local_data_len_or_byte_cap = data_len;
        if (!flags.equal_lens) local_data_len_or_byte_cap = reader.uint32();

        const auto it = state.key_data_types.find(key);
        if (it == state.key_data_types.end())
            throw std::runtime_error("Unknown channel key: " + std::to_string(key));

        auto s = x::telem::Series(it->second, local_data_len_or_byte_cap);
        s.time_range = ref_tr;
        s.alignment = ref_alignment;

        s.fill_from(reader);

        if (!flags.equal_time_ranges) {
            s.time_range.start = x::telem::TimeStamp(reader.int64());
            s.time_range.end = x::telem::TimeStamp(reader.int64());
        }

        if (!flags.equal_alignments) s.alignment = x::telem::Alignment(reader.uint64());

        if (frame.channels == nullptr) {
            frame.channels = std::make_unique<std::vector<channel::Key>>();
            frame.series = std::make_unique<std::vector<x::telem::Series>>();
        }

        frame.emplace(key, std::move(s));
        return !reader.exhausted();
    };

    if (flags.all_channels_present) {
        frame.reserve(state.keys.size());
        for (const auto &key: state.keys)
            decode_series(key);
    } else
        while (decode_series(reader.uint32())) {}

    return {std::move(frame), x::errors::NIL};
}
}
